### PR TITLE
Parametrize stateful failover tests

### PR DESCRIPTION
**Pull request by rosik**
Run the same tests for both stateboard and etcd2.
Reshuffle test code to make it compatible with both cases.
Also fix some flaky bugs:
- call `wish_state` before `apply_config`;
- add few more checks for the sake of synchronization;
- fix mistake in promotion error message;

I didn't forget about

- [x] Tests
- [x] Changelog (unnecessary)
- [x] Documentation (unnecessary)

### CODE REVIEW

**Comment by olegrok** [comment]
> Also fix some bugs

Please list them in commit message

File: test/integration/failover_stateful_test.lua [hunk 1]
```diff
@@ -66,36 +50,155 @@ g.before_all(function()
...
+            '--data-dir', self.workdir,
+            '--listen-peer-urls', 'http://localhost:17001',
+            '--listen-client-urls', URI,
+            '--advertise-client-urls', URI,
+        })
+
+        helpers.retrying({}, function()
+            local resp = httpc.put(URI .. '/v2/keys/hello?value=world')
+            assert(resp.status == 200, resp.body)
+            require('log').info(
```
Comment (olegrok) [inline test/integration/failover_stateful_test.lua:130]: require is redundant

File: test/integration/failover_stateful_test.lua [hunk 2]
```diff
@@ -66,36 +50,155 @@ g.before_all(function()
...
+        end)
+    end
+    function g.state_provider:stop()
+        local process = self.process
+        if process == nil then
+            return
+        end
+        self.process:kill()
+        helpers.retrying({}, function()
+            t.assert_not(process:is_alive(), 'Etcd is still running')
```
Comment (olegrok) [inline test/integration/failover_stateful_test.lua:142]: IMO you shouldn't fail after timeout. Just send SIGKILL if it's still alive.

File: test/integration/failover_stateful_test.lua [hunk 3]
```diff
@@ -66,36 +50,155 @@ g.before_all(function()
...
+        require('log').info('Coordinator options updated')
     ]])
-    g.cluster.main_server.net_box:call(
+end
+
+g_stateboard.before_all(function()
+    local g = g_stateboard
+    g.datadir = fio.tempdir()
+
+    fio.mktree(fio.pathjoin(g.datadir, 'stateboard'))
```
[the remark this answers is not in the document]
Comment (rosik) [reply test/integration/failover_stateful_test.lua]: It's redundant. Removed it completely.

File: test/integration/failover_stateful_test.lua (hunk 1 above)
Comment (rosik) [reply test/integration/failover_stateful_test.lua]: Fixed

File: test/integration/failover_stateful_test.lua (hunk 2 above)
Comment (rosik) [reply test/integration/failover_stateful_test.lua:150]: I don't want to:

1. It's copy-pasted from test helpers
2. There's no convenient way to do it.

File: test/integration/failover_stateful_test.lua [hunk 4]
```diff
@@ -132,107 +246,103 @@ function g.test_stateboard_restart()
         }})
     end)
 
-    g.stateboard:start()
+    fio.rmtree(g.state_provider.workdir)
```
Comment (olegrok) [inline test/integration/failover_stateful_test.lua:249]: Do you really need it? If you drop state_provider data you don't restart it. You boot completely new state_provider instance.

File: test/integration/failover_stateful_test.lua [hunk 5]
```diff
@@ -132,107 +246,103 @@ function g.test_stateboard_restart()
...
-        }
-    )
+end)
 
+add('test_leader_restart', function(g)
     -----------------------------------------------------
-    g.cluster:server('storage-1'):stop()
-    t.assert_equals(
-        g.stateboard.net_box:call('longpoll', {3}),
+    g.client:longpoll(0)
```
Comment (olegrok) [inline test/integration/failover_stateful_test.lua:313]: Why did you add longpoll(0) here?

File: test/integration/failover_stateful_test.lua (hunk 4 above)
Comment (rosik) [reply test/integration/failover_stateful_test.lua:249]: Yes.

1. The intention was to cover the case of clean state provider, as you correctly pointed out.
2. This patch doesn't add anything new. This line originally was above `:stop()`, but sometimes it produces unnecessary warnings in logs, so I moved it here.

File: test/integration/failover_stateful_test.lua (hunk 5 above)
Comment (rosik) [reply test/integration/failover_stateful_test.lua:313]: It's already been there. It's purpose is to abandon all previous events and act as a sync point of S1 death.
